### PR TITLE
Let torch.device be overrideable by TorchFunctionMode

### LINKED ISSUE #103828: simple and small optimization: default device argument for custom device
### 🚀 The feature, motivation and pitch

1、For many operators(such as pin_memory), the device argument is  default as `cuda` if not given; but for other device, we must have to give extra argument device_type comparing to cuda, so we add an API to set the default argument device just once at the begining to keep usage consistent with cuda.
2、And there are some API defined in Python, we add a argument named device_type and the default value is cuda, (such as https://github.com/pytorch/pytorch/blob/main/torch/random.py#L104), so that we could support more device (privateuse1 device).
So we want to add an API to set the default argument device just once at the begining to keep usage consistent with cuda, and add an api to get the default device to keep usage consistent with cuda if not gived device_type.

### Alternatives

_No response_

### Additional context

_No response_

cc @albanD

### PR DESCRIPTION
**Pull request by dilililiwhy**
Fixes #103828
let torch.device be overrideable by TorchFunctionMode

### CODE REVIEW

File: torch/csrc/Device.cpp [hunk 1]
```diff
@@ -267,6 +274,7 @@ void THPDevice_init(PyObject* module) {
     throw python_error();
   }
   Py_INCREF(&THPDeviceType);
+  THPUpperModuleOfDevice = PyImport_ImportModule("torch");
```
Comment (ezyang) [inline torch/csrc/Device.cpp:277]: You don't have to do this, just stash module somewhere durable and use that

**Comment by ezyang** [comment]
Looks good, just one slight modification please!

File: torch/csrc/Device.cpp (hunk 1 above)
Comment (dilililiwhy) [reply torch/csrc/Device.cpp]: Thanks for the review.
Declaration and assignment have merged.

**Review by ezyang — APPROVED** [review]

**Comment by dilililiwhy** [comment]
@ezyang Sorry to bother you, but i have no ideas about failed check **linux-jammy-py3.9-clang12-asan**. 
```
ERROR: ld.so: object '/usr/lib/llvm-7/lib/clang/7.0.1/lib/linux/libclang_rt.asan-x86_64.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
==38==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
Error: Process completed with exit code 1.
```


As i find the `/usr/lib/llvm-12/lib/clang/12.0.1/lib/linux/libclang_rt.asan-x86_64.so` in building stage, maybe the hardcode of llvm and clang version in _.ci/pytorch/test.sh_ (`.../llvm-7/lib/clang/7.0.1/...`) caused above six test failures?

File: torch/csrc/Device.cpp (hunk 1 above)
Comment (dilililiwhy) [reply torch/csrc/Device.cpp]: OK, I have made the change

**Pull request closed without merging**